Added MP Rotator

fern-aviles opened the pull request:
What's new
-
- MP Rotator has been added to the database
- User can enter their desired pressure
- User can now select an available product

Set Up
-
- Clone repo
- In your command line:
  - run npm install to install necessary packages
  - run npm run dev to run the project in your web browser

On your browser, go to localhost:5173 and you should see a page with an input field and select field at the top with a black line and text reading "5 feet" under it.

Testing
-
- Click to add an MP-Rotator
- There should be 18 different nozzles to choose from
- Clicking on a nozzle shows information based on the current water arc and radius and based on the product's recommended pressure
- Previous functionality should remain the same
  - Nozzles options change depending on the arc and radius
  - Nozzles become deselected if its arc or radius don't match the current water arc and radius

Commit:
Added MP Rotator

Commit:
Switched product added on click

Commit:
Added enter PSI and select Product interface

scott-allison commented:
Solve for user story: As a user, I have an arc setting at 240º and radius at 20 ft. I want to increase the radius to 32 ft so that I can cover my landscape area, with as few clicks as possible. - Solve for situation where no nozzles are available because the water object violates all arc constraints at the current radius. Automatically adjust the arc start/end angle to achieve a valid nozzle choice.

scott-allison commented:
Default chosen nozzle: For MP, always pre-select the first available option that is not HT.

scott-allison commented:
Add a Nozzle dropdown with selection next to product selection. The activeSelection should populate the nozzle dropdown.

Commit:
Added nozzle selection dropdown menu

- User can see the available nozzles but only select one that meets the
  requirements for the water setting
- Selecting a different product on the canvas updates the nozzles
  dropdown menu

Commit:
Forgot to upload water.ts

Commit:
Product and water enhancements

- Product changes color depending on nozzle chosen
- Water skips omitted angles

Commit:
Arc auto adjusting added

- If there exists a nozzle with the current radius whos arc is invalid,
  switch to a valid arc while remaining at the same radius

Commit:
Arc auto adjusting improved

Commit:
PGP Ultra Modifications

  - Removed color changing
  - Removed auto selecting

Commit:
Nozzle readjusts data after change

Commit:
Prevent error on setNozzle if argument is empty

Commit:
Changed color names to hex values

scott-allison commented on src/util/product.ts:
```diff
@@ -158,114 +189,197 @@ export class Product extends Circle {
+    const nozzles = data[id].nozzles;
+    let mainMaxArc = 0;
+    for(let nozzle in nozzles){
+      const numbers = nozzles[nozzle].model;
```
Suggest rename const numbers to const model.

scott-allison commented on src/util/product.ts:
```diff
@@ -158,114 +189,197 @@ export class Product extends Circle {
+    let mainMaxArc = 0;
+    for(let nozzle in nozzles){
+      const numbers = nozzles[nozzle].model;
+      for(let number in numbers){
```
model in models

scott-allison commented on src/util/product.ts:
```diff
@@ -158,114 +189,197 @@ export class Product extends Circle {
+    for(let nozzle in nozzles){
+      const numbers = nozzles[nozzle].model;
+      for(let number in numbers){
+        const key = `${nozzle}, ${number}`;
```
change number to model

scott-allison commented on src/util/product.ts:
```diff
@@ -158,114 +189,197 @@ export class Product extends Circle {
+      for(let number in numbers){
+        const key = `${nozzle}, ${number}`;
+
+        const pressures = numbers[number];
```
pressures = models[model];

scott-allison commented on src/util/product.ts:
```diff
@@ -158,114 +189,197 @@ export class Product extends Circle {
-              this.water.canvas.add(this.nozzleOptions[key].text);
-          }
+    for(let nozzle in nozzles){ 
+      const numbers = nozzles[nozzle].model;
```
Change to models here too

scott-allison reviewed (changes requested):
Continuing my review...

Commit:
Fixed auto nozzle selection bugs

- Some nozzles were selectable when they were not supposed to
- MP3500 nozzle wasn't automatically selectable when deselected
- Improved code readability
- Removed unused function
- Added scaling factors to MP Rotators

TODO: Fix bug where the arc rotates when it's at its max arc and the
user continues dragging the arc using the start/end controller

scott-allison commented on src/util/product.ts:
```diff
@@ -158,160 +190,361 @@ export class Product extends Circle {
+    return closestAngle;
+  }
+
+  setNozzle(e: string){ 
```
Add documentation
